pkg/mattermost/mattermost.go

```diff
@@ -0,0 +1,206 @@
+		logging.Logger.Fatal(fmt.Sprintf("Error in loading configuration. Error:%s", err.Error()))
+	}
+
+	return &Bot{
+		ServerURL:    c.Communications.Mattermost.URL,
+		Token:        c.Communications.Mattermost.Token,
+		TeamName:     c.Communications.Mattermost.Team,
+		ChannelName:  c.Communications.Mattermost.Channel,
+		ClusterName:  c.Settings.ClusterName,
+		AllowKubectl: c.Settings.AllowKubectl,
+	}
+}
+
+// Channel structure in Mattermost
+func mmChannel(channelName, teamID string) *model.Channel {
+	return &model.Channel{
+		Name:        channelName,
+		DisplayName: channelName,
+		Purpose:     ChannelPurpose,
+		Type:        model.CHANNEL_OPEN,
+		TeamId:      teamID,
+	}
+}
+
+// Start establishes mattermost connection and listens for messages
+func (b *Bot) Start() {
+	client = model.NewAPIv4Client(b.ServerURL)
+	client.SetOAuthToken(b.Token)
+
+	// Check connection to Mattermost server
+	err := checkServerConnection()
+	if err != nil {
+		logging.Logger.Error("There was a problem pinging the Mattermost server. Error: ", err)
+		return
+	}
+
+	// Check Team exists and get Team ID
+	botTeam, err = getBotTeam(b.TeamName)
+	if err != nil {
+		logging.Logger.Error("There was a problem finding Mattermost team. Error: ", err)
+		return
+	}
+
+	// Check Botkube user exists and get User ID
+	botUser, err = getBotUser(botTeam.Id)
+	if err != nil {
+		logging.Logger.Error("There was a problem creating user in Mattermost. Error: ", err)
+		return
+	}
+
+	// Check Channel exists or create Channel and add user to the Channel
+	botChannel, err = getBotChannel(b.ChannelName, botTeam.Id, botUser.Id)
+	if err != nil {
+		logging.Logger.Error("There was a problem creating channel. Error: ", err)
+		return
+	}
+
+	// Create WebSocketClient and handle messages
+	webSocketURL := WebSocketProtocol + strings.SplitN(b.ServerURL, ":", 2)[1]
+	webSocketClient, _ := model.NewWebSocketClient4(webSocketURL, client.AuthToken)
+	webSocketClient.Listen()
+	go func() {
+		for {
+			select {
+			case event := <-webSocketClient.EventChannel:
+				handleMessage(event, b.AllowKubectl, b.ClusterName, b.ChannelName)
+			}
+		}
+	}()
+	return
+}
+
+// Check if Mattermost server is reachable
+func checkServerConnection() error {
+	if _, resp := client.GetOldClientConfig(""); resp.Error != nil {
+		return resp.Error
+	}
+	return nil
+}
+
+// Check if team exists in Mattermost
+func getBotTeam(teamName string) (*model.Team, error) {
+	botTeam, resp := client.GetTeamByName(teamName, "")
+	if resp.Error != nil {
+		return botTeam, resp.Error
+	}
+	return botTeam, nil
+}
+
+// Check if botkube user exists in Mattermost
+func getBotUser(teamID string) (*model.User, error) {
+	users, resp := client.AutocompleteUsersInTeam(teamID, BotName, "")
+	if resp.Error != nil {
+		return nil, resp.Error
+	}
+	return users.Users[0], nil
+}
+
+// Create channel if not present and add botkube user in channel
+func getBotChannel(channelName, botTeamID, botUserID string) (*model.Channel, error) {
```

Comment:
Can we create MattermostMessage{} struct with all required variables and implement methods on that instead of passing parameters in each function?

Reply:
Sure. Will change and commit.